### PR TITLE
Iteration on Nomad/Variables <-> Nextflow/Secrets

### PR DESCRIPTION
**Pull request by abhi18av**
Use Nomad Variables as Secrets

A user can enable/disable the use of Nomad Variables as Secrets. Nomad Variables are configuration stored in the cluster and used as templates in the job definition. In this way the definition doesn't reveal the values of them

this PR implements a NomadSecretProvider to act as bridge between processes/workflows and the cluster 

Also it implements several commands to get/set/delete/list variables (if the token has enough access)

The majority of design level discussion happened on the https://github.com/nextflow-io/nf-nomad/discussions/73

### CODE REVIEW

File: validation/secrets/main.nf
Comment (abhi18av) [inline validation/secrets/main.nf:6]: As of [tweak the sun-nomadlab config to accommodate variables [ci skip]](https://github.com/nextflow-io/nf-nomad/pull/75/commits/d4e7015702efe622df53bef15cdc96530bb3d35b) , the directive level usage of secrets works as expected. 


However, I suspect we might need to extend the `SecretsLoader` class from Nextflow to be able to use `nomad variables` as secrets in configuration files.

I think we don't need to accommodate multiple sources i.e. `local` + `nomad` in the plugin and can switch completely to `nomad variables` for all secrets.

For example 


```

#NOTE This is stored in sun-nomadlab cluster
aws {
    accessKey = secrets.SUN_NOMADLAB_MINIO_ACCESS_KEY
    secretKey = secrets.SUN_NOMADLAB_MINIO_SECRET_KEY
    client {
        endpoint = 'http://100.119.165.23:9000'
    }
}

```

**Review by abhi18av — COMMENTED** [review]
@jagedn, I've added a comment after a test with the `nf-nomad` secrets at the config level.

Don't think is needs to be a priority but something nice to have and we can address this in the next iteration - whatever you feel best 👍

File: validation/secrets/main.nf
Comment (jagedn) [reply validation/secrets/main.nf:6]: the problem is that current implementation/architecture of Nextflow core doesn't allow custom SecretsLoader to access configurations. This is why our SecretsLoader implementation is a "fake" implementation (it returns null in all get secret requests as all of them will be resolved as nomad templates)

also not sure can mix different SecretsLoader, as far I can see Nextflow expects only 1 active

Comment (abhi18av) [reply validation/secrets/main.nf:6]: Yup, for our purposes I don't particularly see any issues with the `homogenous` secret stores so that's all good.

I guess a true prod cluster, would always work in conjunction with the `vault` setup as @tomiles  also mentioned.

Then, let's move ahead with the current implementation and testing for the next release `v.0.1.2` ?

Comment (jagedn) [reply validation/secrets/main.nf:6]: @abhi18av will last commit I think we cover your comment, let me know if it's not clear

**Review by jagedn — APPROVED** [review]

File: validation/secrets/main.nf
Comment (jagedn) [inline validation/secrets/main.nf:22]: with current implementation we can access to nomad variables also in workflows!!!

**Review by jagedn — APPROVED** [review]

File: validation/secrets/main.nf
Comment (abhi18av) [reply validation/secrets/main.nf:6]: Okay, so I tested the plugin as of ccb80bd9e4b8a1c51eca47b8250a9e9441d601f0 and here are the report.


### ✅ `sun-nomadlab` with local NF secrets store + job vars from Nomad


```
nextflow run -c sun-nomadlab/nextflow.config secrets/main.nf -w 's3://fusionfs/integration-test/work'
```

And this works as expected, by relying completely upon the following config

```
nomad.jobs.secrets.enable = false

aws {
// From local Nextflow secret store
    accessKey = secrets.SUN_NOMADLAB_ACCESS_KEY
    secretKey = secrets.SUN_NOMADLAB_SECRET_KEY

}

```


### 🔴 `sun-nomadlab` with Nomad secret store + job vars from Nomad


In this case, the problem is that 

```
nomad.jobs.secrets.enable = true // OR false


aws {
// From Nomad variables secret store
      accessKey = secrets.SUN_NOMADLAB_MINIO_ACCESS_KEY
      secretKey = secrets.SUN_NOMADLAB_MINIO_SECRET_KEY

}
```

Which raises the following Error, related to unknown variable being used within `aws` scope.


```

ERROR ~ The Access Key Id you provided does not exist in our records. (Service: Amazon S3; Status Code: 403; Error Code: InvalidAccessKeyId; Request ID: 17E69CA81533D4ED; S3 Extended Request ID: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8; Proxy: null)

 -- Check '.nextflow.log' file for details

```

File: plugins/nf-nomad/src/main/nextflow/nomad/config/NomadSecretOpts.groovy
Comment (abhi18av) [inline plugins/nf-nomad/src/main/nextflow/nomad/config/NomadSecretOpts.groovy]: @jagedn , I've renamed the `enable` to `enabled` in order to comply with the general pattern of enabling scopes as per [Nextflow configuration](https://www.nextflow.io/docs/latest/config.html#scope-k8s)

File: validation/secrets/main.nf
Comment (abhi18av) [reply validation/secrets/main.nf:6]: Actually, I must add a clarification here since as of now we have two scenarios


| Setup        | LocalStore | NomadStore |
|--------------|------------|------------|
| Config vars  | ✅          | 🔴          |
| Job vars     | ✅          | ✅          |
| sun-nomadlab | ✅          | ✅          |


### ✅ Scenario-1 

Config vars => LocalSecret store

Job vars    => LocalSecret store



### ✅ Scenario-2 

Config vars => LocalSecret store

Job vars    => NomadSecret store



### 🔴  Scenario-3

Config vars => NomadSecret store

Job vars    => NomadSecret store




### Conclusion

@jagedn , please correct me if I'm wrong - in `scenario-3` we can't use the `NomadSecret` store for `config vars` due to the current design decisions on the Nextflow side, however we can support `scenario-1` and `scenario-2` just fine.

Comment (jagedn) [reply validation/secrets/main.nf:6]: problem here comes because the AWS implementation is not looking for a SecretProvider (I need to investigate a little more):


```
jul-29 09:28:28.700 [main] DEBUG n.c.aws.util.S3CredentialsProvider - No AWS credentials available - falling back to anonymous access
jul-29 09:28:29.528 [main] ERROR nextflow.cli.Launcher - @unknown
com.amazonaws.services.s3.model.AmazonS3Exception: Access Denied. (Service: Amazon S3; Status Code: 403; Error Code: AccessDenied; Request ID: 17E69E6F84261E40; S3 Extended Request ID: dd9025bab4ad464b049177c95eb6ebf374d3b3fd1af9251148b658df7ac2e3e8; Proxy: null)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1879)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1418)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1387)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1157)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:814)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:781)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:755)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:715)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:697)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:561)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:541)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5520)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5467)
	at com.amazonaws.services.s3.AmazonS3Client.access$300(AmazonS3Client.java:422)
	at com.amazonaws.services.s3.AmazonS3Client$PutObjectStrategy.invokeServiceCall(AmazonS3Client.java:6601)
	at com.amazonaws.services.s3.AmazonS3Client.uploadObject(AmazonS3Client.java:1891)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1851)
	at nextflow.cloud.aws.nio.S3Client.putObject(S3Client.java:203)
	at nextflow.cloud.aws.nio.S3FileSystemProvider.createDirectory(S3FileSystemProvider.java:488)
	at java.base/java.nio.file.Files.createDirectory(Files.java:700)
	at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:807)
	at java.base/java.nio.file.Files.createDirectories(Files.java:753)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.selectMethod(IndyInterface.java:355)
	at org.codehaus.groovy.vmplugin.v8.IndyInterface.fromCache(IndyInterface.java:321)
	at nextflow.extension.FilesEx.mkdirs(FilesEx.groovy:493)
	at nextflow.Session.init(Session.groovy:404)
	at nextflow.script.ScriptRunner.execute(ScriptRunner.groovy:129)
	at nextflow.cli.CmdRun.run(CmdRun.groovy:372)
	at nextflow.cli.Launcher.run(Launcher.groovy:503)
	at nextflow.cli.Launcher.main(Launcher.groovy:657)
```

Comment (jagedn) [reply validation/secrets/main.nf:6]: I was debugging and this is the flow I think is happened:

start the pipeline with nomad.secrets.enabled = true (or false it doesnt matter)

core load the plugin and the plugin reset the SecretProvider to have a chance to activate our NomadSecretProvider

core compile the pipeline before to run it. The compilation see a `secrets` code and call to LocalSecretProvider 

once compiled the core load all the plugins and now our SecretProvider is loaded .... but it's to late

I'll try to figure how to fix it

Comment (jagedn) [reply validation/secrets/main.nf:6]: To summarize:

I think you can't use our SecretProvider (nor any other SecretProvider except LocalProvider) to solve secrets values in the config , at least with current version of Nexflow

Comment (abhi18av) [reply validation/secrets/main.nf:6]: > problem here comes because the AWS implementation is not looking for a SecretProvider


Mmm this is interesting, since technically the AWS implementation (or any other config) should be loaded after the secrets are resolved which means, the following works with baseline nextflow 

```groovy
aws {

// From local LocalSecrets store
    accessKey = secrets.SUN_NOMADLAB_ACCESS_KEY
    secretKey = secrets.SUN_NOMADLAB_SECRET_KEY
}

```


> once compiled the core load all the plugins and now our SecretProvider is loaded .... but it's to late


Agreed, so I guess that the mechanism which is used for `plugin` load isn't the same for loading `secrets` i.e. registering and then over-riding previous configs.

> I need to investigate a little more


Actually, @jagedn , I think what we have already is good enough for the `0.1.2` so that we can focus on the testing with real world pipelines and then documentation.

I'm quite happy if we track this use case (`scenario-3`) separately and then address this sometime in future - what do you think? 

Let's go for the release? 😊

Comment (jagedn) [reply validation/secrets/main.nf:6]: Yes, I agree

scenario 3 probably will require some work by core team so .... let's go!!!

Comment (abhi18av) [reply validation/secrets/main.nf:6]: > I think you can't use our SecretProvider (nor any other SecretProvider except LocalProvider) to solve secrets values in the config , at least with current version of Nexflow

Agreed.

So maybe we should revert/stash the changes related to the `NomadSecrets` implementation so that our `plugin cmd` commands can work again as expected. As of now, it seems to run into problems during plugin resolution 🤔 

```
$ nextflow -c sun-nomadlab/nextflow.config plugin nf-nomad:secrets list       

Invalid target plugin: nf-nomad
```

Comment (jagedn) [reply validation/secrets/main.nf:6]: If I'm not wrong will be automagically resolved once we publish the version in the plugin repository 

I mean, right now the plugin cmd is trying to use last version from repository and it doesnt implement cmd

Comment (abhi18av) [reply validation/secrets/main.nf:6]: Ah, okay - then let's merge-tag-pr 😎

Comment (abhi18av) [reply validation/secrets/main.nf:6]: After the merge to `plugins` repo https://github.com/nextflow-io/plugins/pull/83#event-13679519573 , I can confirm that the `nf-nomad:secrets` commands are working as expected ✅ 

image

**Pull request merged**
